istioctl/pkg/rbac/upgrade.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rbac
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"istio.io/istio/pilot/pkg/config/kube/crd"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/istio/pilot/pkg/model"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+)
+
+// ServiceToWorkloadLabels maps the short service name to the workload labels that it's pointing to.
+// This service is defined in same namespace as the ServiceRole that's using it.
+type ServiceToWorkloadLabels map[string]WorkloadLabels
+
+// WorkloadLabels is the workload labels, for example, app: productpage.
+type WorkloadLabels map[string]string
+
+type Upgrader struct {
+	IstioConfigStore model.ConfigStore
+	K8sClient        *kubernetes.Clientset
+	// RoleToWorkloadLabels maps the ServiceRole name to the map of service name to the workload labels
+	// that the service pointing to.
+	// We could have remove the `service` key layer to make it ServiceRole name maps to workload labels.
+	// However, this is needed for unit tests.
+	RoleToWorkloadLabels map[string]ServiceToWorkloadLabels
+	RbacFile             string
+	serviceRoles         []model.Config
+	serviceRoleBindings  []model.Config
+}
+
+const (
+	creationTimestampNilField = "creationTimestamp: null"
+)
+
+var (
+	configDescriptor = model.ConfigDescriptor{
+		model.ServiceRole,
+		model.ServiceRoleBinding,
+		model.AuthorizationPolicy,
+	}
+)
+
+// UpgradeCRDs is the main function that converts RBAC v1 to v2 for local policy files.
+func (ug *Upgrader) UpgradeCRDs() (string, error) {
+	err := ug.createRoleAndBindingLists(ug.RbacFile)
+	var convertedPolicies strings.Builder
+	if err != nil {
+		return "", err
+	}
+	for _, serviceRole := range ug.serviceRoles {
+		if err := ug.updateRoleToWorkloadLabels(serviceRole); err != nil {
+			return "", err
+		}
+		role := ug.upgradeServiceRole(serviceRole)
```

Comment:
I think we should just restrict to conversion for now, i.e. not checking if policy is correct or not, or if policy is not used. Later on we can create another sub command, e.g. `istioctl experimental rbac check rbacfile.yaml` to check the things I mentioned above. This reduces the burden for the rbac conversion tool (like a micro service vs a monolithic application :D). In the future we can bundle the two features if we think it's a better UX.

Reply:
The error checking should be part of the `upgrade` command, we should make it error-tolerant at some level. 

For example, ServiceRole A in fileA is not referred by any ServiceRoleBinding. Then it's better to ignore the ServiceRole A instead of generating a ServiceRole A without service field. I'm okay to do this in a follow-up PR but please add a TODO for this.

Reply:
Will do in a follow-up PR.